src/AI/WallyAI.java

```diff
@@ -0,0 +1,788 @@
+              {
+                log(String.format("  Gonna try to kill %s, who has %s HP", target.toStringWithLocation(), target.getHP()));
+                double damageSum = 0;
+                for( XYCoord xyc : neededAttacks.keySet() )
+                {
+                  Unit unit = neededAttacks.get(xyc);
+                  if( null != unit )
+                  {
+                    damageSum += CombatEngine.simulateBattleResults(unit, target, gameMap, xyc.xCoord, xyc.yCoord).defenderHPLoss;
+                    actions.offer(new GameAction.AttackAction(gameMap, unit, Utils.findShortestPath(unit, xyc, gameMap), target.x, target.y));
+                    unitQueue.remove(unit);
+                    log(String.format("    %s brings the damage total to %s", unit.toStringWithLocation(), damageSum));
+                    if (damageSum >= target.getPreciseHP())
+                      break;
+                  }
+                }
+                stateChange = true;
+                foundKill = true;
+                break;
+              }
+              else
+              {
+                // log(String.format("  Can't kill %s, oh well", target.toStringWithLocation()));
+              }
+            }
+          }
+          if( foundKill )
+            break;
+        }
+      }
+
+      // Figure out where we don't wanna go
+      Map<UnitModel, Map<XYCoord, Double>> threatMap = new HashMap<UnitModel, Map<XYCoord, Double>>();
+      if( actions.isEmpty() )
+      {
+        Map<Commander, ArrayList<Unit>> unitLists = AIUtils.getEnemyUnitsByCommander(myCo, gameMap);
+        for( UnitModel um : myCo.unitModels )
+        {
+          threatMap.put(um, new HashMap<XYCoord, Double>());
+          for( Commander co : unitLists.keySet() )
+          {
+            if( myCo.isEnemy(co) )
+            {
+              for( Unit threat : unitLists.get(co) )
+              {
+                // add each new threat to the existing threats
+                Map<XYCoord, Double> threatArea = threatMap.get(um);
+                for( Entry<XYCoord, Double> newThreat : AIUtils.findThreatPower(gameMap, threat, um).entrySet() )
+                {
+                  if( null == threatArea.get(newThreat.getKey()) )
+                    threatArea.put(newThreat.getKey(), newThreat.getValue());
+                  else
+                    threatArea.put(newThreat.getKey(), newThreat.getValue() + threatArea.get(newThreat.getKey()));
+                }
+              }
+            }
+          }
+        }
+      }
+
+      // try to get unit value by capture or attack
+      while (actions.isEmpty() && !unitQueue.isEmpty())
+      {
+        Unit unit = unitQueue.poll();
+        XYCoord position = new XYCoord(unit.x, unit.y);
+        
+        if (unit.getCaptureProgress() > 0)
+        {
+          actions.offer(new GameAction.CaptureAction(gameMap, unit, Utils.findShortestPath(unit, position, gameMap)));
+          capturingProperties.add(position);
```

Comment:
Based on the logic in `initTurn`, `position` should already be a member of `capturingProperties` is there any reason to add it twice? Alternatively, perhaps `capturingProperties` should be a set?

Reply:
Hah, good catch.